src/SignalR/server/Core/src/Internal/ChannelExtensions.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+internal static class ChannelExtensions
+{
+    public static ValueTask RunAsync<TState>(this ChannelBasedSemaphore channelSemaphore, Func<TState, Task> callback, TState state)
+    {
+        if (channelSemaphore.AttemptWait())
```

Comment:
`AttemptWait` should probably be private (and probably called `TryAcquire` because this makes no attempt whatsoever to actually wait lol) and these methods can move to ChannelBasedSemaphore. I don't see the benefit of extension methods.